### ISSUE TITLE
AsyncEnumerator.MoveNextAsync extension method considered harmful?

### ISSUE THREAD

**Issue by Tragetaschen** [dotnet/reactive#2175]
#### Problem
We have run at least twice into issues by accidentally using the [AsyncEnumerator.MoveNextAsync](https://github.com/dotnet/reactive/blob/fefe75993de542bc8dc6e01287fce0956a6770ab/Ix.NET/Source/System.Linq.Async/System/Linq/AsyncEnumerator.cs#L51-L59) extension method and I have trouble finding a scenario where it would be applicable.

> Which subcomponent library (Ix, Async.Ix)?

Async.Ix

> Which library version?

6.0.1

> What is the problem?

1) The extension method is very easy to discover and use, it lives in a namespace that is part of the default global usings nowadays.
2) It speaks to a correctness argument: "Yes, of course I want cancellation when I call that async method".
3) During code review everything looks perfectly fine.
4) It doesn't do what you can reasonably expect. The cancellation token is only synchronously checked when entering the method, but never used along the async chain of the underlying async enumerator.

We have run into issues where the async enumerator was created without cancellation token and the expectation was that calling MoveNextAsync would cancel. However, the resulting code would only cancel when the async enumerator made forward progress and re-entered the next MoveNextAsync call. That didn't happen and we were left with a deadlock or memory leak.

> What is the expected outcome?

I would vote for removing / deprecating this extension method, but I also realize that this is a breaking change.

I have put in the global namespace another overload of that extension method with the same signature and ObsoleteAttribute, so it is always preferred by the compiler and disables my own usage.

**Comment by julealgon** [comment]
Why does this method exist in the first place, considering `IAsyncEnumerable` now exists as a first-class citizen in the framework?

**Comment by idg10** [comment]
`System.Linq.Async` has now effectively been deprecated by .NET's own `System.Linq.AsyncEnumerable`. If you use v7 of `System.Linq.Async` it brings in a transitive dependency on `System.Linq.AsyncEnumerable`, and in all cases where functionality previously supplied by `System.Linq.Async` is now built into `System.Linq.AsyncEnumerable`, we've removed it from the public API of `System.Linq.Async`, and you will end up using the new implementation in `System.Linq.AsyncEnumerable` even if you are not referencing that directly.

We will be deprecating `System.Linq.Async` soon. The only reason to continue to use it is if you need things it contains that are not in .NET's own `System.Linq.AsyncEnumerable` and which have not been relocated to `System.Interactive.Async`. All of the functionality that is a) still not in the .NET runtime libraries and b) plausibly useful, now lives in `System.Interactive.Async`. (For example, if you look at the `AsyncEnumerator` class being discussed here, there is perhaps some use for its `Create` method (even though you're almost certainly better off using language-level iterator support) so we've reproduced that as `AsyncEnumeratorEx.Create` in `System.Interactive.Async`. But you'll see that `AsyncEnumeratorEx` does **not** have `MoveNextAsync`.)

So the only reason to continue to use `System.Linq.Async` is either because you require binary compatibility (e.g., you've got a dependency on some component that was built against `System.Linq.Async` v6 or older, which expects to find certain methods in `System.Linq.Async.dll`), or your own code still uses some deprecated feature. (E.g., perhaps someone is using this non-useful `MoveNextAsync`. They should definitely stop because it's not useful, but we don't want to completely break their build.)

In short, starting with the release of .NET 10 and the corresponding release of Ix.NET v7, the only point of `System.Linq.Async` is to continue to provide the old API for backwards compatibility purposes. (And to clarify the situation for .NET FX: the runtime libraries' new `System.Linq.AsyncEnumerable` that shipped as part of .NET 10.0 can also be used from .NET FX.)

So although this particular method might have been an unfortunate design choice (or possibly just an incomplete implementation) we're not going to remove it, because the only reason anybody should continue to use `System.Linq.Async` is because they continue to have a dependency on some of the old methods it defined that were not brought forward into either `System.Linq.AsyncEnumerable` or `System.Interactive.Async`.

I.e., methods like this problematic one are now in fact the only reason for `System.Linq.Async` to continue to exist!

Thus, we can't remove it. Instead, we're asking people to stop using `System.Linq.Async`.

Should we mark it as deprecated? I had been thinking that since we were going to deprecate the whole NuGet package, the implication is that everything in it is deprecated. However, you've got me wondering whether it's worth also marking all the remaining methods as `Obsolete`. Obviously that would help right now since we've not yet marked the package as deprecated. (We wanted to give v7 a few months of use in practice to find out whether we'd got anything disastrously wrong. It was surprisingly tricky to make it fit around the new `System.Linq.AsyncEnumerable`—we had to use various tricks involving making the reference assemblies present a different API surface area from the runtime DLLs, and we were worried that this might have unintended consequences requiring a major rethink, and a hasty v7.1 release. So far, fortunately, we don't seem to have any major new problems there.)

In the longer term, one scenario that could be improved my deprecating this (and perhaps some other lingering back-compat-only methods) is when some project is not yet able to rid itself of a dependency on `System.Linq.Async`. (E.g., perhaps you have a dependency on some other component that you need, but which has not yet been updated to remove its dependency on `System.Linq.Async`.) In these cases, the remaining public API surface area of `System.Linq.Async` is now available to your project, but you really don't want to be using it, because your long term goal should be to stop using `System.Linq.Async`. So marking all its public methods as `Obsolete` might be a good idea, so that you get warned if you accidentally use something like this that looks like the right thing, but which turns out to be doomed.


@julealgon wrote:
> Why does this method exist in the first place, considering IAsyncEnumerable now exists as a first-class citizen in the framework?

I don't know the full history because this predates my involvement in Rx. However, in the first place (like, back to 2012) `IAsyncEnumerable` wasn't even defined by the framework: it was an invention of the Rx.NET team, and before .NET Core 3 came out (in 2019), the `IAsyncEnumerable<T>` interface itself was defined in an Ix.NET package.

When `IAsyncEnumerable<T>` was added to the framework in .NET Core 3, Ix.NET was modified so that it no longer defined its own version of this interface. However, `IAsyncEnumerable<T>` was still not a first-class citizen: at that time, Microsoft did not see fit to implement a LINQ provider for it. So instead, the Ix.NET implementation `System.Linq.Async` became the de facto LINQ for `IAsyncEnumerable<T>`.

So from 2012 through to 2025, `IAsyncEnumerable<T>` was not a first-class citizen in the framework. It looks like this particular method was added around 2018, which I expect was in anticipation of the `await foreach` language feature, and and it appears like it was meant to support cancellable iteration. But it does look like it was half-baked. It makes me wonder if the intention was to make it work in the way you would expect, but that the ball was dropped, and this never happened.

It was only with the release of .NET 10.0 that `IAsyncEnumerable<T>` finally became a first-class citizen. And it was at that point that we released a Ix.NET v7, which basically steps out of .NET's way here. The intention is that nobody should continue to use `System.Linq.Async`. The features that component used to offer that are still not provided in .NET 10 are now in `System.Interactive.Async`, and we did **not** include `MoveNextAsync` in that set of relocated features.

**Comment by idg10** [comment]
TLDR version of the above:

* the `AsyncEnumerator.MoveNextAsync` method is effectively deprecated because it exists only in `System.Linq.Async`
* `System.Linq.Async` has not yet been officially marked as a deprecated package but will be soon (because the .NET runtime library `System.Linq.AsyncEnumerable` package effectively replaces Ix.NET's de-facto-but-not-really-official LINQ for `IAsyncEnumerable<T>`)
* even so, I'm now considering marking the public methods that remain in `System.Linq.Async` as `[Obsolete]` to raise awareness that these are things you should stop using
